rust/main/agents/relayer/src/server/merkle_tree_insertions/insert_merkle_tree_insertions.rs

```diff
@@ -0,0 +1,335 @@
+use std::collections::HashMap;
+
+use axum::{extract::State, http::StatusCode, routing::post, Json, Router};
+use derive_new::new;
+use hyperlane_base::db::HyperlaneRocksDB;
+use hyperlane_core::{MerkleTreeInsertion, H256};
+use serde::{Deserialize, Serialize};
+
+use crate::server::utils::{ServerErrorResponse, ServerResult, ServerSuccessResponse};
+
+#[derive(Clone, Debug, new)]
+pub struct ServerState {
+    pub dbs: HashMap<u32, HyperlaneRocksDB>,
+}
+
+impl ServerState {
+    pub fn router(self) -> Router {
+        Router::new()
+            .route("/merkle_tree_insertions", post(handler))
+            .with_state(self)
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct TreeInsertion {
+    pub chain: u32,
+    pub insertion_block_number: u64,
+
+    pub leaf_index: u32,
+    pub message_id: H256,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct RequestBody {
+    pub merkle_tree_insertions: Vec<TreeInsertion>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ResponseBody {
+    pub count: u64,
+}
```

Comment:
I think it would be useful if we add information about what records were not stored into database if user misprinted the domain.